### PR TITLE
Access Error Messages of OnboardingService and RevokeService

### PR DESCRIPTION
**Pull request by Frank-Wiebeler**
+ADD: Adding possibility to receive error message that leads to the OnboardingService throwing an error.
-DEL: Removed duplicate of Protobuf Definitions import

### CODE REVIEW

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/RevokingService.java [hunk 1]
```diff
@@ -1,9 +1,15 @@
...
 import com.dke.data.agrirouter.api.service.parameters.RevokeParameters;
+import java.util.Optional;
 
 public interface RevokingService {
 
   RevokeResponse revoke(RevokeParameters revokeParameters);
+
+  String getLastErrorAsString();
+
+  Optional<RevokingError> getLastError();
```
Comment (saschadoemer) [inline agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/RevokingService.java]: I would prefer `getRevokingError` and `getLastError` to avoid `...asString` within the method name.

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java [hunk 2]
```diff
@@ -16,4 +18,10 @@
   OnboardingResponse onboard(OnboardingParameters parameters);
 
   String generateAuthorizationUrl(AuthorizationRequestParameters parameters);
+
+  /** @return The last error, if the last onboarding failed as a String */
+  String getLastErrorAsString();
```
Comment (saschadoemer) [inline agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java]: Should be valid JavaDoc. `return` contains the description.

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java [hunk 3]
```diff
@@ -16,4 +18,10 @@
...
+
+  /** @return The last error, if the last onboarding failed as object read from the JSON String */
+  Optional<OnboardingError> getLastError();
```
Comment (saschadoemer) [inline agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java]: Should be valid JavaDoc. `return` contains the description.

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/secured/OnboardingService.java [hunk 4]
```diff
@@ -29,4 +31,10 @@
    * @return -
    */
   String generateAuthorizationUrl(AuthorizationRequestParameters parameters);
+
+  /** @return The last error, if the last onboarding failed */
+  String getLastErrorAsString();
+
+  /** @return The last error, if the last onboarding failed as object read from the JSON String */
+  Optional<OnboardingError> getLastError();
```
Comment (saschadoemer) [inline agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/secured/OnboardingService.java]: Same here, I would prefer `getOnboardingError` and `getLastError` as field accessor.

File: agrirouter-api-java-api/src/test/java/com/dke/data/agrirouter/api/factories/impl/OnboardErrorTest.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import com.dke.data.agrirouter.api.dto.onboard.OnboardingError;
+import com.google.gson.Gson;
+import org.junit.jupiter.api.Test;
+
+public class OnboardErrorTest {
+  @Test
+  public void readError() {
```
Comment (saschadoemer) [inline agrirouter-api-java-api/src/test/java/com/dke/data/agrirouter/api/factories/impl/OnboardErrorTest.java]: Pattern for test methods should be `given ... do something ... should` for example `givenResultAsJsonParseErrorShouldReturnObjectContainingValidErrorCode`

File: agrirouter-api-java-api/src/test/java/com/dke/data/agrirouter/api/factories/impl/RevokeErrorTest.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import com.dke.data.agrirouter.api.dto.revoke.RevokingError;
+import com.google.gson.Gson;
+import org.junit.jupiter.api.Test;
+
+public class RevokeErrorTest {
+  @Test
+  public void readRevokeError() {
```
Comment (saschadoemer) [inline agrirouter-api-java-api/src/test/java/com/dke/data/agrirouter/api/factories/impl/RevokeErrorTest.java]: See other comment.

File: agrirouter-api-java-convenience/src/main/java/com/dke/data/agrirouter/convenience/mqtt/client/MqttOptionService.java [hunk 7]
```diff
@@ -62,7 +62,8 @@ public MqttConnectOptions createMqttConnectOptions(OnboardingResponse onboarding
   private SocketFactory getSocketFactory(
       String certificate, String password, CertificationType certificationType) {
     try {
-      KeyManagerFactory kmf = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+      KeyManagerFactory kmf =
```
Comment (saschadoemer) [inline agrirouter-api-java-convenience/src/main/java/com/dke/data/agrirouter/convenience/mqtt/client/MqttOptionService.java]: Fomatter issues?

File: agrirouter-api-java-impl/src/main/java/com/dke/data/agrirouter/impl/onboard/OnboardingServiceImpl.java [hunk 8]
```diff
@@ -49,15 +54,23 @@ private OnboardingRequest createOnboardingRequestBody(OnboardingParameters param
...
     Response response =
         RequestFactory.bearerTokenRequest(this.environment.getOnboardUrl(), registrationCode)
             .post(Entity.json(onboardingRequest));
-    this.assertStatusCodeIsCreated(response.getStatus());
-    OnboardingResponse onboardingResponse = response.readEntity(OnboardingResponse.class);
+    try {
+      response.bufferEntity();
+      this.lastError = response.readEntity(String.class);
+      this.assertStatusCodeIsCreated(response.getStatus());
+      this.lastError = "";
```
Comment (saschadoemer) [inline agrirouter-api-java-impl/src/main/java/com/dke/data/agrirouter/impl/onboard/OnboardingServiceImpl.java]: This leads to confusion. Setting the error to reset it later is bad style. Perhaps it is better to put the error - if there is one - into the exception to handle it all at one place.

File: agrirouter-api-java-impl/src/main/java/com/dke/data/agrirouter/impl/onboard/secured/OnboardingServiceImpl.java [hunk 9]
```diff
@@ -152,4 +163,20 @@ public String generateAuthorizationUrl(AuthorizationRequestParameters parameters
...
   }
+
+  @Override
+  public String getLastErrorAsString() {
+    return this.lastError;
+  }
+
+  @Override
+  public Optional<OnboardingError> getLastError() {
+    if (this.lastError == null || this.lastError.equals("")) {
```
Comment (saschadoemer) [inline agrirouter-api-java-impl/src/main/java/com/dke/data/agrirouter/impl/onboard/secured/OnboardingServiceImpl.java]: Use `StringUtils` for it - could also just contain whitespaces...

File: pom.xml [hunk 10]
```diff
@@ -144,14 +144,6 @@
                 <version>${project.version}</version>
             </dependency>
 
-            <!-- PROTOBUF API -->
```
Comment (saschadoemer) [inline pom.xml]: Why did you remove these lines?

Comment (Frank-Wiebeler) [reply pom.xml]: They are a duplicate of Line 128-133, but in difference to 128-133, JDK8 is hardcoded. 
It lead to several maven issues.

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/RevokingService.java (hunk 1 above)
Comment (Frank-Wiebeler) [reply agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/RevokingService.java]: Done

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java (hunk 2 above)
Comment (Frank-Wiebeler) [reply agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java]: Done

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/secured/OnboardingService.java (hunk 4 above)
Comment (Frank-Wiebeler) [reply agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/secured/OnboardingService.java]: Done

File: agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java (hunk 3 above)
Comment (Frank-Wiebeler) [reply agrirouter-api-java-api/src/main/java/com/dke/data/agrirouter/api/service/onboard/OnboardingService.java]: Done

File: agrirouter-api-java-api/src/test/java/com/dke/data/agrirouter/api/factories/impl/OnboardErrorTest.java (hunk 5 above)
Comment (Frank-Wiebeler) [reply agrirouter-api-java-api/src/test/java/com/dke/data/agrirouter/api/factories/impl/OnboardErrorTest.java]: Removed, as it only "Tested itself". Maybe add a different test in the future; would require a setLastError-Function though

File: agrirouter-api-java-convenience/src/main/java/com/dke/data/agrirouter/convenience/mqtt/client/MqttOptionService.java (hunk 7 above)
Comment (Frank-Wiebeler) [reply agrirouter-api-java-convenience/src/main/java/com/dke/data/agrirouter/convenience/mqtt/client/MqttOptionService.java]: Should have been corrected by mvn clean install, however, fixed it by hand.

Comment (Frank-Wiebeler) [reply agrirouter-api-java-convenience/src/main/java/com/dke/data/agrirouter/convenience/mqtt/client/MqttOptionService.java]: It's formatted like that if I do mvn clean install, so, the "problem" is at least part of the project

File: agrirouter-api-java-impl/src/main/java/com/dke/data/agrirouter/impl/onboard/OnboardingServiceImpl.java (hunk 8 above)
Comment (Frank-Wiebeler) [reply agrirouter-api-java-impl/src/main/java/com/dke/data/agrirouter/impl/onboard/OnboardingServiceImpl.java]: Done here, done with revoke, done with secured
